// @webspatial/builder@1.6.0/dist/lib/utils/sceneUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultSceneConfigVolume = exports.defaultSceneConfig = exports.SpatialSceneValues = exports.WorldAlignmentValues = exports.WorldScalingValues = exports.BaseplateVisibilityValues = void 0;
exports.isValidBaseplateVisibilityType = isValidBaseplateVisibilityType;
exports.isValidWorldScalingType = isValidWorldScalingType;
exports.isValidWorldAlignmentType = isValidWorldAlignmentType;
exports.isValidSpatialSceneType = isValidSpatialSceneType;
exports.formatToNumber = formatToNumber;
exports.formatSceneConfig = formatSceneConfig;
exports.isValidSceneUnit = isValidSceneUnit;
exports.BaseplateVisibilityValues = [
    'automatic',
    'visible',
    'hidden',
];
function isValidBaseplateVisibilityType(type) {
    return exports.BaseplateVisibilityValues.includes(type);
}
exports.WorldScalingValues = ['automatic', 'dynamic'];
function isValidWorldScalingType(type) {
    return exports.WorldScalingValues.includes(type);
}
exports.WorldAlignmentValues = [
    'adaptive',
    'automatic',
    'gravityAligned',
];
function isValidWorldAlignmentType(type) {
    return exports.WorldAlignmentValues.includes(type);
}
exports.SpatialSceneValues = ['window', 'volume'];
function isValidSpatialSceneType(type) {
    return exports.SpatialSceneValues.includes(type);
}
function pxToMeter(px) {
    return px / 1360;
}
function meterToPx(meter) {
    return meter * 1360;
}
function formatToNumber(str, targetUnit, defaultUnit) {
    if (typeof str === 'number') {
        if ((defaultUnit === 'px' && targetUnit === 'px') ||
            (defaultUnit === 'm' && targetUnit === 'm')) {
            return str;
        }
        // unit not match target
        if (defaultUnit === 'px' && targetUnit === 'm') {
            return pxToMeter(str);
        }
        else if (defaultUnit === 'm' && targetUnit === 'px') {
            return meterToPx(str);
        }
        // fallback
        return str;
    }
    if (targetUnit === 'm') {
        if (str.endsWith('m')) {
            // 1m
            return Number(str.slice(0, -1));
        }
        else if (str.endsWith('px')) {
            // 100px
            return pxToMeter(Number(str.slice(0, -2)));
        }
        else {
            throw new Error('formatToNumber: invalid str');
        }
    }
    else if (targetUnit === 'px') {
        if (str.endsWith('px')) {
            // 100px
            return Number(str.slice(0, -2));
        }
        else if (str.endsWith('m')) {
            // 1m
            return meterToPx(Number(str.slice(0, -1)));
        }
        else {
            throw new Error('formatToNumber: invalid str');
        }
    }
    else {
        throw new Error('formatToNumber: invalid targetUnit');
    }
}
function formatSceneConfig(config, sceneType) {
    // defaultSize and resizability's width/height/depth can be 100 or "100px" or "1m"
    // expect:
    // resizability should format into px
    // defaultSize should format into px if window
    // defaultSize should format into m if volume
    const isWindow = sceneType === 'window';
    // const isVolume = sceneType === 'volume'
    // format resizability
    if (config.resizability) {
        const iterKeys = Object.keys(config.resizability);
        for (let k of iterKeys) {
            if (config.resizability[k]) {
                ;
                config.resizability[k] = formatToNumber(config.resizability[k], 'px', isWindow ? 'px' : 'm');
            }
        }
    }
    // format defaultSize
    if (config.defaultSize) {
        const iterKeys = Object.keys(config.defaultSize);
        for (let k of iterKeys) {
            if (config.defaultSize[k]) {
                ;
                config.defaultSize[k] = formatToNumber(config.defaultSize[k], isWindow ? 'px' : 'm', isWindow ? 'px' : 'm');
            }
        }
    }
    return config;
}
/**
 * check px,m and number, number must be >= 0
 *
 * */
function isValidSceneUnit(val) {
    // only support number or string with unit px or m
    // rpx cm mm not allowed
    if (typeof val === 'number') {
        return val >= 0;
    }
    if (typeof val === 'string') {
        if (val.endsWith('px')) {
            // check if number
            if (isNaN(Number(val.slice(0, -2)))) {
                return false;
            }
            return Number(val.slice(0, -2)) >= 0;
        }
        if (val.endsWith('m')) {
            // check if number
            if (isNaN(Number(val.slice(0, -1)))) {
                return false;
            }
            return Number(val.slice(0, -1)) >= 0;
        }
    }
    return false;
}
exports.defaultSceneConfig = {
    defaultSize: {
        width: 1280,
        height: 720,
        depth: 0,
    },
};
exports.defaultSceneConfigVolume = {
    defaultSize: {
        width: 0.94,
        height: 0.94,
        depth: 0.94,
    },
};
