// @webspatial/builder@1.6.0/dist/lib/resource/imageHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImageHelper = void 0;
const Jimp = require("jimp");
const sharp = require("sharp");
const resvg_js_1 = require("@resvg/resvg-js");
class ImageHelper {
    static createImg(size) {
        return new Jimp(size, size, 0x00000000);
    }
    static async webp2PngBuffer(buffer) {
        return await sharp(buffer).toFormat('png').toBuffer();
    }
    static async svg2img(svg) {
        const opt = {
            fitTo: {
                mode: 'width',
                value: 1024, // Generate the SVG with 1024px width, for larger icons.
            },
        };
        const resvg = new resvg_js_1.Resvg(svg, opt);
        const pngData = resvg.render();
        const pngBuffer = pngData.asPng();
        return pngBuffer;
    }
    static isFullyOpaque(image) {
        const pixelNum = image.getWidth() * image.getHeight();
        for (var i = 0; i < pixelNum; i++) {
            const idx = i * 4 + 3;
            if (image.bitmap.data[idx] < 255) {
                return false;
            }
        }
        return true;
    }
}
exports.ImageHelper = ImageHelper;
